velox/connectors/hive/SplitReader.cpp

```diff
@@ -250,23 +252,47 @@ void SplitReader::setNullConstantValue(
 namespace {
 
 template <TypeKind ToKind>
-velox::variant convertFromString(const std::optional<std::string>& value) {
+velox::variant convertFromString(
+    const std::optional<std::string>& value,
+    const TypePtr& toType,
+    memory::MemoryPool* pool,
+    core::ExpressionEvaluator* expressionEvaluator) {
+  using T = typename TypeTraits<ToKind>::NativeType;
   if (value.has_value()) {
     if constexpr (ToKind == TypeKind::VARCHAR) {
       return velox::variant(value.value());
     }
     if constexpr (ToKind == TypeKind::VARBINARY) {
       return velox::variant::binary((value.value()));
     }
-    auto result = velox::util::Converter<ToKind>::cast(value.value());
     if constexpr (ToKind == TypeKind::TIMESTAMP) {
+      auto result = velox::util::Converter<ToKind>::cast(value.value());
       result.toGMT(Timestamp::defaultTimezone());
+      return velox::variant(result);
     }
-    return velox::variant(result);
+
+    // Prepare input row vector.
+    std::vector<VectorPtr> children = {
+        BaseVector::createConstant(VARCHAR(), value.value(), 1, pool)};
+    const auto input = std::make_shared<RowVector>(
+        pool, ROW({"c0"}, {VARCHAR()}), BufferPtr(nullptr), 1, children);
+
+    // Create cast expression to convert string to toType.
+    const core::TypedExprPtr inputField =
+        std::make_shared<const core::FieldAccessTypedExpr>(VARCHAR(), "c0");
+    const core::TypedExprPtr callExpr =
+        std::make_shared<const core::CallTypedExpr>(
+            toType, std::vector<core::TypedExprPtr>{inputField}, "cast");
+
+    const std::unique_ptr<exec::ExprSet> exprSet =
+        expressionEvaluator->compile(callExpr);
+    VectorPtr result;
+    const SelectivityVector rows(1);
+    expressionEvaluator->evaluate(exprSet.get(), rows, *input, result);
```

Comment:
Using expression evaluation here feels a bit heavy weight. Why can't we use a lower-level utility? Do we not have access to query config?

Reply:
Thanks for your review. These was a deprecated query config `isIso8601`, but could not be leaked to connector as previous discussion https://github.com/facebookincubator/velox/pull/7084#discussion_r1379279568 pointed out. Do you think it makes sense to add a configuration in HiveConfig? BTW, the semantics of cast string as timestamp are also different between Presto and Spark. Do we intend to use configurations to solve them?

Reply:
> Why can't we use a lower-level utility?

Hi @mbasmanova, I took a further look here. The format of partitionKeys can be controlled by Gluten, so if we ensure that compatible strings are generated as partition values, maybe we don't need to bring more complexity here. For date format `YYYY-MM-DD`, `castFromDateString` with true for `isIso8601` should be enough. Do you think it makes sense?